flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/TestFlinkUpsert.java

```diff
@@ -0,0 +1,288 @@
+              "PARTITIONED BY (data) WITH %s",
+          tableName, toWithClause(tableUpsertProps));
+
+      sql("INSERT INTO %s VALUES " +
+          "(1, 'aaa')," +
+          "(2, 'aaa')," +
+          "(3, 'bbb')",
+          tableName);
+
+      List<Row> result = sql("SELECT * FROM %s", tableName);
+
+      Assert.assertEquals("result should have the correct rows",
+          Sets.newHashSet(Row.of(2, "aaa"), Row.of(3, "bbb")), Sets.newHashSet(result));
+
+      sql("INSERT INTO %s VALUES " +
+          "(4, 'aaa')," +
+          "(5, 'bbb')",
+          tableName);
+
+      result = sql("SELECT * FROM %s", tableName);
+
+      Assert.assertEquals("result should have the correct rows",
+          Sets.newHashSet(Row.of(4, "aaa"), Row.of(5, "bbb")), Sets.newHashSet(result));
+
+      sql("INSERT INTO %s VALUES " +
+          "(6, 'aaa')," +
+          "(7, 'bbb')",
+          tableName);
+
+      result = sql("SELECT * FROM %s", tableName);
+
+      Assert.assertEquals("result should have the correct rows",
+          Sets.newHashSet(Row.of(6, "aaa"), Row.of(7, "bbb")), Sets.newHashSet(result));
+    } finally {
+      sql("DROP TABLE IF EXISTS %s.%s", flinkDatabase, tableName);
+    }
+  }
+
+  @Test
+  public void testCompoundPrimaryKey() {
+    String tableName = "upsert_primary_compound_key";
+    LocalDate dt = LocalDate.of(2022, 3, 1);
+    try {
+      sql("CREATE TABLE %s(id INT, data STRING NOT NULL, dt DATE NOT NULL, PRIMARY KEY(data,dt) NOT ENFORCED) " +
+              "PARTITIONED BY (data) WITH %s",
+          tableName, toWithClause(tableUpsertProps));
+
+      sql("INSERT INTO %s VALUES " +
+          "(1, 'aaa', TO_DATE('2022-03-01'))," +
+          "(2, 'aaa', TO_DATE('2022-03-01'))," +
+          "(3, 'bbb', TO_DATE('2022-03-01'))",
+          tableName);
+
+      List<Row> result = sql("SELECT * FROM %s", tableName);
+
+      Assert.assertEquals("result should have the correct rows",
+          Sets.newHashSet(Row.of(2, "aaa", dt), Row.of(3, "bbb", dt)), Sets.newHashSet(result));
+
+      sql("INSERT INTO %s VALUES " +
+          "(4, 'aaa', TO_DATE('2022-03-01'))," +
+          "(5, 'bbb', TO_DATE('2022-03-01'))",
+          tableName);
+
+      result = sql("SELECT * FROM %s", tableName);
+
+      Assert.assertEquals("result should have the correct rows",
+          Sets.newHashSet(Row.of(4, "aaa", dt), Row.of(5, "bbb", dt)), Sets.newHashSet(result));
+
+      sql("INSERT INTO %s VALUES " +
+          "(6, 'aaa', TO_DATE('2022-03-01'))," +
+          "(7, 'bbb', TO_DATE('2022-03-01'))",
+          tableName);
+
+      result = sql("SELECT * FROM %s", tableName);
+
+      Assert.assertEquals("result should have the correct rows",
+          Sets.newHashSet(Row.of(6, "aaa", dt), Row.of(7, "bbb", dt)), Sets.newHashSet(result));
+    } finally {
+      sql("DROP TABLE IF EXISTS %s.%s", flinkDatabase, tableName);
+    }
+  }
+
+  @Test
+  public void testPrimaryKeyFieldsAtEndOfTableSchema() {
```

Comment:
are the method name and comment below accurate? primary key fields (data and dt) seem located at the beginning of the schema.

Reply:
Good catch. Let me fix that.

Reply:
Updated these!